webapp/channels/src/packages/mattermost-redux/src/reducers/entities/channel_categories.ts

```diff
@@ -92,6 +95,44 @@ export function byId(state: IDMappedObjects<ChannelCategory> = {}, action: Gener
 
         return changed ? nextState : state;
     }
+    case ChannelTypes.RECEIVED_CHANNEL: {
+        const receivedChannel: Channel = toClientChannel(action.data);
+        if (receivedChannel.type === General.PRIVATE_CHANNEL) {
+            // For GM to Private channel conversion feature
+            // In the case when someone converts your GM to a private channel and moves it to a team
+            // you're not currently on, we need to remove the channel from "direct messages" category
+            // and add it to "channels" category of target team. Even though the server sends a websocket event about updated category data,
+            // it does so only for the team the channel got moved into, and not every team a user is part of, for performance reasons.
+            // For every other team, we update the state here. Everything is correct ion server side, but we update state here
+            // to avoid re-fetching all categories again for all teams.
+
+            const newState: IDMappedObjects<ChannelCategory> = {};
+            const categoryIDs = Object.keys(state);
+            categoryIDs.forEach((categoryID) => {
+                if (categoryID.startsWith('direct_messages_')) {
+                    // for DM category, remove the new channel from the category.
+                    newState[categoryID] = {
+                        ...state[categoryID],
+                        channel_ids: state[categoryID].channel_ids.filter((channelID) => channelID !== receivedChannel.id),
+                    };
+                } else if (categoryID.startsWith('channels_') && state[categoryID].team_id === receivedChannel.team_id && state[categoryID].channel_ids.indexOf(receivedChannel.id) < 0) {
+                    // We don't need to worry about adding the channel in the right order as this is only
+                    // an intermediate step, meant to handle teh edge case of missing the upcoming "update category"
+                    // websocket message, triggered on conversion of GM to private channel.
+                    newState[categoryID] = {
+                        ...state[categoryID],
+                        channel_ids: [...state[categoryID].channel_ids, receivedChannel.id],
+                    };
+                } else {
+                    newState[categoryID] = state[categoryID];
+                }
+            });
+
+            return newState;
```

Comment:
Relating to my above comment, we could do something like this, but we'd want to make sure that it only modifies the state when there's actually something to change. Since this isn't an array that we can use `.reduce` on, the best way to do that would probably be to keep track of if we actually made any changes like how `LEAVE_CHANNEL` is handled in an earlier case and then we end this with returning `changed ? newState : state`

Reply:
Handled in new code. In new code, I'm updating category state only when its a conversion by check if channel already exists in state as a GM and new channel is a private channel.